c/xudt_rce.c

```diff
@@ -0,0 +1,526 @@
+    // overflows.
+    if (input_amount < current_amount) {
+      return ERROR_OVERFLOWING;
+    }
+    i += 1;
+  }
+
+  // With the sum of all input UDT tokens gathered, let's now iterate through
+  // output cells to grab the sum of all output UDT tokens.
+  uint128_t output_amount = 0;
+  i = 0;
+  while (1) {
+    uint128_t current_amount = 0;
+    len = 16;
+    // Similar to the above code piece, we are also looping through output cells
+    // with the same script as current running script here by using
+    // `CKB_SOURCE_GROUP_OUTPUT`.
+    ret = ckb_checked_load_cell_data((uint8_t*)&current_amount, &len, 0, i,
+                                     CKB_SOURCE_GROUP_OUTPUT);
+    if (ret == CKB_INDEX_OUT_OF_BOUND) {
+      break;
+    }
+    if (ret != CKB_SUCCESS) {
+      return ret;
+    }
+    if (len < 16) {
+      return ERROR_ENCODING;
+    }
+    output_amount += current_amount;
+    // Like any serious smart contract out there, we will need to check for
+    // overflows.
+    if (output_amount < current_amount) {
+      return ERROR_OVERFLOWING;
+    }
+    i += 1;
+  }
+
+  // When both value are gathered, we can perform the final check here to
+  // prevent non-authorized token issurance.
+  if (input_amount < output_amount) {
+    return ERROR_AMOUNT;
+  }
+  return CKB_SUCCESS;
+}
+
+// If the extension script is identical to a lock script of one input cell in
+// current transaction, we consider the extension script to be already
+// validated, no additional check is needed for current extension
+int is_extension_script_validated(mol_seg_t extension_script,
+                                  uint8_t* input_lock_script_hash,
+                                  uint32_t input_lock_script_hash_count) {
+  int err = 0;
+  uint8_t hash[BLAKE2B_BLOCK_SIZE];
+  err = blake2b(hash, BLAKE2B_BLOCK_SIZE, extension_script.ptr,
+                extension_script.size, NULL, 0);
+  CHECK2(err == 0, ERROR_BLAKE2B_ERROR);
+
+  for (uint32_t i = 0; i < input_lock_script_hash_count; i++) {
+    if (memcmp(&input_lock_script_hash[i * BLAKE2B_BLOCK_SIZE], hash,
+               BLAKE2B_BLOCK_SIZE) == 0) {
+      return 0;
+    }
+  }
+  err = ERROR_NOT_VALIDATED;
+exit:
+  return err;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main() {
+#endif
+  // don't move code_buff into global variable. It doesn't work.
+  // it's a ckb-vm bug: the global variable will be freezed:
+  // https://github.com/nervosnetwork/ckb-vm/blob/d43f58d6bf8cc6210721fdcdb6e5ecba513ade0c/src/machine/elf_adaptor.rs#L28-L32
+  // The code can't be loaded into freezed memory.
+  uint8_t code_buff[MAX_CODE_SIZE] __attribute__((aligned(RISCV_PGSIZE)));
+  uint32_t code_used = 0;
+
+  int err = 0;
+  int owner_mode = 0;
+  uint8_t* raw_extension_data = NULL;
+  uint32_t raw_extension_len = 0;
+  XUDTFlags flags = XUDTFlagsPlain;
+  uint8_t
+      input_lock_script_hash[MAX_LOCK_SCRIPT_HASH_COUNT * BLAKE2B_BLOCKBYTES];
```

Comment:
`BLAKE2B_BLOCKBYTES` is defined in blake2b.h which value is 128, must be a typo here. I think the right value is `BLAKE2B_BLOCK_SIZE`.

Reply:
`input_lock_script_hashes` would be a better variable name.